crates/iota/src/signing.rs

```diff
@@ -0,0 +1,138 @@
+    let sender = &tx_data.sender();
+
+    let signature = {
+        let key = context.config().keystore().get_key(sender)?;
+
+        match key {
+            StoredKey::KeyPair(_) => context.config().keystore().sign_secure(
+                sender,
+                tx_data,
+                Intent::iota_transaction(),
+            )?,
+            StoredKey::External {
+                derivation_path,
+                source,
+                ..
+            } => {
+                match ExternalKeySource::from(source.as_str()) {
+                    ExternalKeySource::Ledger => {
+                        if let Some(derivation_path) = derivation_path {
+                            let signer = LedgerSigner::new_with_default(
+                                derivation_path.clone(),
+                                Some(iota_client.clone()),
+                            )?;
+                            // pass the transaction sender to the signer to ensure the correct
+                            // key is used
+                            signer
+                                .sign_transaction(tx_data, sender)
+                                .await
+                                .map(|s| s.signature)?
+                        } else {
+                            bail!(
+                                "Derivation path is required for Ledger signing. Please specify it in the keystore."
+                            );
+                        }
+                    }
+                    ExternalKeySource::Unknown(name) => {
+                        bail!("External signing is not supported for source: {name}")
+                    }
+                }
+            }
+        }
+    };
+    Ok(signature)
+}
+
+pub(crate) fn sign_secure<T>(
+    keystore: &impl AccountKeystore,
+    address: &IotaAddress,
+    msg: &T,
+    intent: Intent,
+) -> Result<Signature>
+where
+    T: Serialize,
+{
+    let key = keystore.get_key(address)?;
+    let iota_signature = match key {
+        StoredKey::KeyPair(_) => keystore.sign_secure(address, &msg, intent)?,
+        StoredKey::External {
+            derivation_path,
+            source,
+            ..
+        } => {
+            match ExternalKeySource::from(source.as_str()) {
+                ExternalKeySource::Ledger => {
+                    if let Some(derivation_path) = derivation_path {
+                        let ledger = Ledger::new_with_default()?;
+                        // Pass the expected address to the ledger to ensure the signature is for
+                        // the correct address.
+                        ledger
+                            .sign_intent(derivation_path, address, intent, &msg, vec![])?
+                            .signature
+                    } else {
+                        bail!(
+                            "Derivation path is required for Ledger signing. Please specify it in the keystore."
+                        );
+                    }
```

Comment:
Writing it like this is cleaner I think
```suggestion
                    let Some(derivation_path) = derivation_path else {
                        bail!(
                            "Derivation path is required for Ledger signing. Please specify it in the keystore."
                        );
                    }
                    let ledger = Ledger::new_with_default()?;
                    // Pass the expected address to the ledger to ensure the signature is for
                    // the correct address.
                    ledger
                        .sign_intent(derivation_path, address, intent, &msg, vec![])?
                        .signature
```

Reply:
Same in sign_transaction()